### PR TITLE
TASK: Decouple content object proxy from node data

### PR DESCRIPTION
**Pull request by dfeyer**
This change remove the logic to handle content object proxy from the NodeData and create a dedicated service `ContentObjectProxyService`. This service listen to Doctrine postUpdate event and Node::emitNodePropertyChanged signal from the CR to keep data in sync between the CR and the Doctrine entities.

This is a first step to have a better and cleaner support of ContentObjectProxy in Neos:
# Todos
- [x] Decouple the logic from NodeData
- [x] CLI controller to sync doctrine -> CR
- [ ] The default behaviour should continue to work if we want this feature non breaking
# Nice to have
- [x] Test integration with Redirection packages (generate all redirect when one node is edited
- [ ] More flexible `ContentObjectProxyService` to handle custom logic per proxied classed (maybe based on TypeConverter)
- [ ] Add configuration section in NodeTypes.yaml to configuration the mapping between doctrine entity property and node property.

### CODE REVIEW

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php [hunk 1]
```diff
@@ -1439,6 +1440,39 @@ public function findNodesByRelatedEntities($relationMap)
     }
 
     /**
+     * @param string $identifier
+     * @return \Generator
+     */
+    public function findByContentObjectProxy($identifier)
```
Comment (dfeyer) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php]: I'm fine to move this one to the `ContentProxyableRepository`

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/ContentProxyableRepository.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+use Doctrine\ORM\Internal\Hydration\IterableResult;
+use Doctrine\ORM\QueryBuilder;
+use TYPO3\Flow\Annotations as Flow;
+use TYPO3\TYPO3CR\Exception;
+
+/**
+ * ContentProxyableRepository
+ *
+ * @Flow\Scope("singleton")
+ * @api
```
Comment (dfeyer) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/ContentProxyableRepository.php:24]: Not sure if this repository need to be marked as API

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Service/ContentObjectProxyService.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    protected $synchronizationDisabled = false;
+
+    /**
+     * @param NodeInterface $node
+     * @param string $propertyName name of the property that has been changed/added
+     * @param mixed $oldValue the property value before it was changed or NULL if the property is new
+     * @param mixed $newValue the new property value
+     * @return void
+     */
+    public function setProperty(NodeInterface $node, $propertyName, $oldValue, $newValue)
```
Comment (dfeyer) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Service/ContentObjectProxyService.php:64]: The logic of this method should be move to `DefaultContentObjectHandler` other handler should be configuration in the package setting for each proxied classes or maybe something based on the property mapper

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Service/ContentObjectProxyService.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            $this->persistenceManager->persistAll(true);
+        });
+    }
+
+    /**
+     * Synchronize doctrine properties to the content repository on post update
+     *
+     * @param LifecycleEventArgs $eventArgs
+     */
+    public function postUpdate(LifecycleEventArgs $eventArgs)
```
Comment (dfeyer) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Service/ContentObjectProxyService.php:85]: The logic of this method should be move to `DefaultContentObjectHandler` other handler should be configuration in the package setting for each proxied classes or maybe something based on the property mapper

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Service/ContentProxyableEntityService.php [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+use TYPO3\TYPO3CR\Domain\Model\NodeInterface;
+use TYPO3\TYPO3CR\Domain\Repository\ContentProxyableRepository;
+use TYPO3\TYPO3CR\Domain\Repository\NodeDataRepository;
+use TYPO3\TYPO3CR\Exception;
+
+/**
+ * Manage synchronization between the content repository and doctrine entities
+ *
+ * @Flow\Scope("singleton")
+ * @api
```
Comment (dfeyer) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Service/ContentProxyableEntityService.php:33]: Maybe this one should not be market API

**Comment by dfeyer** [comment]
The CLI output

CLI output

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php [hunk 6]
```diff
@@ -1439,6 +1440,32 @@ public function findNodesByRelatedEntities($relationMap)
     }
 
     /**
+     * @param string $identifier
+     * @param Workspace $workspace
+     * @return array
+     */
+    public function findByContentObjectProxy($identifier, Workspace $workspace)
```
Comment (kitsunet) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php]: What about dimensions?

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Service/ContentProxyableEntityService.php (hunk 5 above)
Comment (kitsunet) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Service/ContentProxyableEntityService.php:33]: Probably not

Comment (kitsunet) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Service/ContentProxyableEntityService.php:33]: Might also need an interface at some point or strategies depending on entity. Think if we add CQRS/ES for entities or someone stores them not in doctrine.

**Pull request closed without merging**